File: 2024/java/day1/CompareLists.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.lang.String;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class CompareLists {

    private static Long differenceOfLists(List<Integer> arr1, List<Integer> arr2) {
        Long total = (long) 0;
        Collections.sort(arr1);
        Collections.sort(arr2);
        for (int i=0; i < arr1.size(); i++) {
            total += Math.abs(arr1.get(i) - arr2.get(i));
        }
        return total;
    }

    private static Long similarityOfLists(List<Integer> arr1, List<Integer> arr2) {
        Long score = (long) 0;
        Set<Integer> set1 = new HashSet<>();
        for (int a : arr1) {
            set1.add(a);
        }

        // count occurences 
        Map<Integer, Integer> counts1 = new HashMap<>();
        for (int e : arr1) {
            if (set1.contains(e)) {
                counts1.put(e, counts1.getOrDefault(e, 0) + 1);
            }
        }
        Map<Integer, Integer> counts2 = new HashMap<>();
        for (int e : arr2) {
            if (set1.contains(e)) {
                counts2.put(e, counts2.getOrDefault(e, 0) + 1);
            }
        }

        // similarity score formula
        for (int a : set1) {
            score += a * counts1.get(a) * counts2.getOrDefault(a, 0);
        }
        return score;
    }

    public static void main(String[] args) {
        String fileName = "input.txt";
        int i = 0;
        List<Integer> arr1 = new ArrayList<>();
        List<Integer> arr2 = new ArrayList<>();

        try (BufferedReader br = new BufferedReader(new FileReader(fileName))) {
            String line;
            while ((line = br.readLine()) != null) {
                i += 1;
                String[] parts = line.trim().split("\\s+"); // handle multiple spaces
                if (parts.length == 2) {
                    int[] nums = Arrays.stream(parts)
                                .mapToInt(Integer::parseInt)
                                .toArray();
                    arr1.add(nums[0]);
                    arr2.add(nums[1]);
                }
            }
            Long ans1 = differenceOfLists(arr1, arr2);
            System.out.println("Difference of lists: " + ans1);

            Long ans2 = similarityOfLists(arr1, arr2);
            System.out.println("Similarity of lists: " + ans2);
                                        
        } catch (RuntimeException e) {
            System.out.println("Error found in line: " + i + " -> " + e.getMessage());
        } catch (IOException e) {
            System.out.println("Problem reading file: " + fileName + "-> " + e.getMessage());
        }
    }               
}

File: 2024/node/day1/comparelists.js
const fs = require('fs');
const readline = require('readline');

function sumDifferences(list1, list2) {
    // find differences between elements of sorted lists
    list1.sort((a, b) => a - b);
    list2.sort((a, b) => a - b);

    let totalDifference = 0;
    for (let i = 0; i < list1.length; i++) {
        totalDifference += Math.abs(list1[i] - list2[i]);
    }
    return totalDifference;
}

function similarityOfLists(list1, list2) {
    // compute similarity score
    // score contributed by each distinct element e: (value of e) * (occurrences of e in list1) * (occurrences of e in list2)
    const freq1 = {};
    const freq2 = {};

    // frequency map for list1
    list1.forEach(num => {
        freq1[num] = (freq1[num] || 0) + 1;
    });
    // frequency map for list2
    list2.forEach(num => {
        freq2[num] = (freq2[num] || 0) + 1;
    });

    let similarityScore = 0;
    for (const num in freq1) {
        if (freq2[num]) {
            similarityScore += num * freq1[num] * freq2[num];
        }
    }
    return similarityScore;
}

function compareLists(filePath) {
    // stream, interface from file
    const fileStream = fs.createReadStream(filePath);
    const rl = readline.createInterface({
        input: fileStream,
        output: process.stdout,
        terminal: false
    });

    // data structures
    const list1 = [];
    const list2 = [];

    // Listen for each line and process it
    rl.on('line', (line) => {
        // Split the line by whitespace and convert the parts to integers
        const [num1, num2] = line.trim().split(/\s+/).map(Number);
        // push onto sortable lists
        list1.push(num1);
        list2.push(num2);
    });
    
    // Handle the end of the file
    rl.on('close', () => {
        // console.log('Finished reading the file');
        // part 1
        const totalDifference = sumDifferences(list1, list2);
        console.log(`Total difference between lists ${totalDifference}`);
        // part 2
        const similarityScore = similarityOfLists(list1, list2);
        console.log(`Similarity score between lists is: ${similarityScore}`);
    });

    // Handle errors
    rl.on('error', (err) => {
        console.error('Error reading the file:', err);
    });
}


const filePath = 'input.txt';
compareLists(filePath);